### ISSUE TITLE
Incorrectly reported parse error

### ISSUE THREAD

**Issue by JLedelay** [utwente-fmt/vercors#875]
The following input:
```java
class C {
  final D d;


  public void m() {
    //@ ghost seq<int> is = [t: int];
    for (int i = 0; i < 10; i++) {
      int j = d.get(i);
      if (true) {
        //@ ghost is = i : is;
      }
    }
  }
}

class D {
  int get(int i);
}
```
Gives the following output:
```
======================================
At Test.java:8:20:
--------------------------------------
    6      //@ ghost seq<int> is = [t: int];
    7      for (int i = 0; i < 10; i++) {
                         [-
    8        int j = d.get(i);
                          -]
    9        if (true) {
   10          //@ ghost is = i : is;
--------------------------------------
Parsing failed: missing ';' at '('
======================================
```
Note that the actual parse error should occur at `//@ ghost is = i : is;`, as the correct syntax for prepending a sequence is `i :: is`. 

Removing the `if` statement, `i` parameter from the `get` method, or the `get` call causes this error to be reported correctly.
If a `;` is added at the suggested location (resulting in `int j = d.get;(i);`), the error is also reported correctly.

**Comment by niomaster** [comment]
Interesting! ANTLR does report some more information than we use, but in principle this is "just because of antlr".

**Comment by niomaster** [comment]
tl;dr we did not collect all errors, but we do now, let's leave it at that for now.

Longer explanation: the full list of errors is as such:

```
======================================
 At examples/private/joel5.java:8:26:
--------------------------------------
    6          //@ ghost seq<int> is = [t: int];
    7          for (int i = 0; i < 10; i++) {
                               [-
    8              int j = d.get(i);
                                -]
    9              if (true) {
   10                  //@ ghost is = i : is;
--------------------------------------
[1/3] Parsing failed: missing ';' at '('
--------------------------------------
 At examples/private/joel5.java:10:34:
--------------------------------------
    8              int j = d.get(i);
    9              if (true) {
                                       [-
   10                  //@ ghost is = i : is;
                                        -]
   11              }
   12          }
--------------------------------------
[2/3] Parsing failed: mismatched input ':' expecting {'instanceof', '[', ';', '.', '=', '>', '<', '?', '==', '<=', '>=', '!=', '&&', '||', '++', '--', '+', '-', '*', '/', '&', '|', '^', '%', '+=', '-=', '*=', '/=', '&=', '|=', '^=', '%=', '<<=', '>>=', '>>>=', '->', BlockStartSpecImmediate, 'then', '\in', '?.', '::', '\', '**', '==>', '-*', BlockStartSpec, LineStartSpec}
--------------------------------------
 At examples/private/joel5.java:10:39:
--------------------------------------
    8              int j = d.get(i);
    9              if (true) {
                                            [-
   10                  //@ ghost is = i : is;
                                             -]
   11              }
   12          }
--------------------------------------
[3/3] Parsing failed: extraneous input '\n' expecting {'inline', VAL_ASSERT, 'abstract', 'assert', 'boolean', 'break', 'byte', 'char', 'class', 'continue', 'do', 'double', 'enum', 'final', 'float', 'for', 'if', 'int', 'interface', 'long', 'new', 'package', 'private', 'protected', 'public', 'return', 'short', 'static', 'strictfp', 'super', 'switch', 'synchronized', 'this', 'throw', 'try', 'void', 'while', IntegerLiteral, FloatingPointLiteral, 'true', 'false', CharacterLiteral, StringLiteral, 'null', '(', '{', '}', '[', ';', '<', '!', '~', '++', '--', '+', '-', '*', '|', '@', BlockStartSpecImmediate, LANG_ID_ESCAPE, 'resource', 'process', 'frac', 'zfrac', 'bool', 'ref', 'rational', 'seq', 'set', 'bag', 'pointer', 'map', 'option', 'either', 'tuple', 'type', 'any', 'nothing', 'pure', 'thread_local', 'with', 'then', 'given', 'yields', 'axiom', 'model', 'adt', 'modifies', 'accessible', 'requires', 'ensures', 'context_everywhere', 'context', 'loop_invariant', 'kernel_invariant', 'lock_invariant', 'signals', 'decreases', 'apply', 'fold', 'unfold', 'open', 'close', 'assume', 'inhale', 'exhale', 'label', 'refute', 'witness', 'ghost', 'send', 'to', 'recv', 'from', 'transfer', 'csl_subject', 'spec_ignore', 'action', 'atomic', 'commit', 'Reducible', 'AddsTo', 'APerm', 'ArrayPerm', 'Contribution', 'held', 'committed', 'HPerm', 'idle', 'perm', 'Perm', 'PointsTo', 'running', 'Some', 'Left', 'Right', 'Value', '\unfolding', '\Unfolding', '\current_thread', '\sum', '\old', '\typeof', '\type', '\matrix', '\array', '\pointer', '\pointer_index', '\pointer_block_length', '\pointer_block_offset', '\pointer_length', '\shared_mem_size', '\values', '\vcmp', '\vrep', '\msum', '\mcmp', '\mrep', '\result', '\ltid', '\gtid', 'none', 'None', 'write', 'read', 'empty', TRIGGER_OPEN, 'b{', '[t:', '{t:', 'b{t:', Identifier, BlockStartSpec, LineStartSpec, VAL_CREATE}
======================================
```

It seems that because [field derefence has higher priority than instance method invocation](https://github.com/utwente-fmt/vercors/blob/c6a2eb347de9e4d2fc5050452f19e3453267b9fb/parsers/lib/antlr4/LangJavaParser.g4#L591-L598) in the java grammar, the parser has to backtrack at some point. What I don't understand is why fixing a seemingly unrelated problem makes the prior error disappear. I think it's probably related to the fact that the `ErrorStrategy` of the parser is involved in both trying to (diagonstically) recover the parser in case of parsing errors (like in this case) and actually reporting errors to the end user. In any case, I'd like to leave it at this for now, since at least the correct error is reported.
